omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+        drvUser->optimisationFlag = "optimisation failed"; // Could not find an optimisation so this tag will be processed directly
+        drvUser->readFlag = true;
+        break;
+      }
+
+      for (auto masterStruct : structIDList)
+      {
+        if (commonStruct.first == masterStruct.first)
+        {
+          // we have found a tag that already exists which gets this struct, so we piggy back off this
+          // now we get the drvUser for the asyn parameter which is to be optimised and set its tagIndex to the tagIndex we piggy back off
+          matchFound = true;
+          destroyList.push_back(drvUser->tagIndex);
+          //std::cout<<"Old value was: "<<drvUser->tagIndex<<std::endl;
+          drvUser->tagIndex = masterStruct.second;
+          //std::cout<<"New value is: "<< this->tagMap_.find(asynIndex)->second->tagIndex<<std::endl;
+          asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Asyn index: %d Optimised to use libplctag tag with libplctag index: %d\n", driverName, functionName, asynIndex, masterStruct.second);
+          if (drvUser->optimisationFlag != "master") {
+            // If we are the master then our tagIndex will be read, otherwise we will rely on the data read by the master         
+            drvUser->optimisationFlag = "optimised";
+            drvUser->readFlag = false;
+          }
+        }
+      }
+      if (!matchFound && commonStruct.second.size() >= countNeeded)
+      {
+        asynPrint(pasynUserSelf, ASYN_TRACE_WARNING, "%s:%s  Warning, Attempt to optimise asyn index: %d failed. The driver will use a default setup.\n", driverName, functionName, asynIndex);
+        drvUser->optimisationFlag = "optimisation failed";
+      }
+    }
+  }
+
+  std::sort(destroyList.begin(), destroyList.end());
+  std::vector<int>::iterator it = std::unique(destroyList.begin(), destroyList.end());
+  destroyList.erase(it, destroyList.end());
+  for (auto index: destroyList) {
+    status = plc_tag_destroy(index); // Destroy old tags
+    asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Destroyed redundant libplctag: %d\n", driverName, functionName, index);
+    if (status!=0) {
+      asynPrint(pasynUserSelf, ASYN_TRACE_WARNING, "%s:%s  Warning, libplctag tag destruction process failed for libplctag index %d\n", driverName, functionName, index);
+    }
+  }
+
+  for (auto tag: tagMap_)
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_FLOW, "%s:%s Asyn param: %d optimisationFlag: '%s' readFlag: %s\n", driverName, functionName, tag.first, tag.second->optimisationFlag.c_str(), tag.second->readFlag ? "true" : "false");
+  }
+
+  this->unlock();
+  this->startPollers_ = true;
+  return asynSuccess; // always return success as any failures to optimise should leave the driver in its pre omtimised state which should have no erros
+}
+
+std::unordered_map<std::string, std::string> drvOmronEIP::drvInfoParser(const char *drvInfo)
```

Comment:
This is a *really* long function and kind of hard to follow from end-to-end. However, it does seem to have different sections. Consider breaking these sections down into smaller (private) functions?

Reply:
Maybe even create a smaller, stand-alone (unit-testable) class that just has the job to parse drvInfo strings...

Reply:
I will refactor this and some other "non driver" related functions in a seperate branch and I do intend to have unit tests for them.